server/http/restapi/call.go

```diff
@@ -5,22 +5,93 @@ import (
 
 	"github.com/pkg/errors"
 
+	pluginapi "github.com/mattermost/mattermost-plugin-api"
+
 	"github.com/mattermost/mattermost-plugin-apps/apps"
 	"github.com/mattermost/mattermost-plugin-apps/server/utils"
 	"github.com/mattermost/mattermost-plugin-apps/server/utils/httputils"
 )
 
-func (a *restapi) handleCall(w http.ResponseWriter, req *http.Request) {
+func (a *restapi) handleCall(w http.ResponseWriter, req *http.Request, sessionID, actingUserID string) {
 	call, err := apps.CallRequestFromJSONReader(req.Body)
 	if err != nil {
 		httputils.WriteError(w, utils.NewInvalidError(errors.Wrap(err, "failed to unmarshal Call request")))
 		return
 	}
 
-	res := a.proxy.Call(sessionID(req), actingID(req), call)
-	if res.Type == "" {
-		res.Type = apps.CallResponseTypeOK
+	cc, err := cleanUserCallContext(a.mm, actingUserID, call.Context)
+	if err != nil {
+		httputils.WriteError(w, utils.NewInvalidError(errors.Wrap(err, "invalid call context for user")))
+		return
 	}
 
+	cc = a.conf.GetConfig().SetContextDefaults(cc)
+
+	call.Context = cc
+	res := a.proxy.Call(sessionID, actingUserID, call)
 	httputils.WriteJSON(w, res)
 }
+
+func cleanUserCallContext(mm *pluginapi.Client, userID string, cc *apps.Context) (*apps.Context, error) {
+	cc = &apps.Context{
+		ContextFromUserAgent: cc.ContextFromUserAgent,
+	}
+
+	var postID, channelID, teamID string
+
+	switch {
+	case cc.PostID != "":
+		postID = cc.PostID
+
+		post, err := mm.Post.GetPost(postID)
+		if err != nil {
+			return nil, err
+		}
+
+		channelID = post.ChannelId
+
+		_, err = mm.Channel.GetMember(channelID, userID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel membership. user=%v channel=%v", userID, channelID)
+		}
+
+		c, err := mm.Channel.Get(channelID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel. channel=%v", channelID)
+		}
+
+		teamID = c.TeamId
+
+	case cc.ChannelID != "":
+		channelID = cc.ChannelID
+
+		_, err := mm.Channel.GetMember(cc.ChannelID, userID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel membership. user=%v channel=%v", userID, cc.ChannelID)
+		}
+
+		c, err := mm.Channel.Get(channelID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel. channel=%v", channelID)
+		}
+
+		teamID = c.TeamId
```

Comment:
Do we need to do that? are inconsistent `cc.ChannelID` and `cc.TeamID` something the apps plugin need to catch?

Reply:
This continues the discussion from https://github.com/mattermost/mattermost-plugin-apps/pull/163#discussion_r618512481. I'm wondering if we are fine with data inconsistencies coming from the user agent, which would grant us a performance boost, because we can drop this call.

Reply:
>  are inconsistent cc.ChannelID and cc.TeamID something the apps plugin need to catch?

@hanzei I think so. A given app should be able to trust that all data passed in is correct and valid

> I'm wondering if we are fine with data inconsistencies coming from the user agent

Maybe we should fail instead of filling in the gaps of the inconsistent values.

Reply:
Discuss further and handle in another PR?